src/buf.rs

```diff
@@ -0,0 +1,637 @@
+        unsafe {
+            let s = slice.as_ptr() as *const grpc_slice as *const UnsafeCell<grpc_slice>;
+            // hack: see From<&GrpcSlice>.
+            GrpcByteBuffer(grpc_raw_byte_buffer_create((*s).get(), len))
+        }
+    }
+}
+
+impl<'a> From<&'a GrpcSlice> for GrpcByteBuffer {
+    /// Create a buffer from the given single slice.
+    ///
+    /// A buffer, which length is 1, is allocated for the slice.
+    #[allow(clippy::cast_ref_to_mut)]
+    fn from(s: &'a GrpcSlice) -> GrpcByteBuffer {
+        unsafe {
+            // hack: buffer_create accepts an mutable pointer to indicate it mutate
+            // ref count. Ref count is recorded by atomic variable, which is considered
+            // `Sync` in rust. This is an interesting difference in what is *mutable*
+            // between C++ and rust.
+            // Using `UnsafeCell` to avoid raw cast, which is UB.
+            let s = &*(s as *const GrpcSlice as *const grpc_slice as *const UnsafeCell<grpc_slice>);
+            GrpcByteBuffer(grpc_raw_byte_buffer_create((*s).get(), 1))
+        }
+    }
+}
+
+impl Clone for GrpcByteBuffer {
+    fn clone(&self) -> Self {
+        unsafe { GrpcByteBuffer(grpc_byte_buffer_copy(self.0)) }
+    }
+}
+
+impl Drop for GrpcByteBuffer {
+    fn drop(&mut self) {
+        unsafe { grpc_byte_buffer_destroy(self.0) }
+    }
+}
+
+/// A zero-copy reader for the message payload.
+///
+/// To achieve zero-copy, use the BufRead API `fill_buf` and `consume`
+/// to operate the reader.
+#[repr(C)]
+pub struct GrpcByteBufferReader {
+    reader: grpc_byte_buffer_reader,
+    /// Current reading buffer.
+    // This is a temporary buffer that may need to be dropped before every
+    // iteration. So use `ManuallyDrop` to control the behavior more clean
+    // and precisely.
+    slice: ManuallyDrop<GrpcSlice>,
+    /// The offset of `slice` that has not been read.
+    offset: usize,
+    /// How many bytes pending for reading.
+    remain: usize,
+}
+
+impl GrpcByteBufferReader {
+    /// Creates a reader for the `GrpcByteBuffer`.
+    ///
+    /// `buf` is stored inside the reader, and dropped when the reader is dropped.
+    pub fn new(buf: GrpcByteBuffer) -> GrpcByteBufferReader {
+        let mut reader = MaybeUninit::uninit();
+        let mut s = MaybeUninit::uninit();
+        unsafe {
+            let code = grpc_byte_buffer_reader_init(reader.as_mut_ptr(), buf.0);
+            assert_eq!(code, 1);
+            if 0 == grpc_byte_buffer_reader_next(reader.as_mut_ptr(), s.as_mut_ptr()) {
+                s.as_mut_ptr().write(grpc_empty_slice());
+            }
+            let remain = grpc_byte_buffer_length((*reader.as_mut_ptr()).buffer_out);
+            // buf is stored inside `reader` as `buffer_in`, so do not drop it.
+            mem::forget(buf);
+
+            GrpcByteBufferReader {
+                reader: reader.assume_init(),
+                slice: ManuallyDrop::new(GrpcSlice(s.assume_init())),
```

Comment:
Could `grpc_empty_slice()` be uninitialized?

Reply:
Better to add a test case for this.

Reply:
The purpose of `grpc_empty_slice` is to create an empty slice, so it has to be initialized. It's guaranteed by the grpc c core.

Reply:
Got it.